File: Front-End/src/app/core/services/vehiculo.service.ts
/**
 * Vehiculo Service - Comunicación con API de vehículos
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiService } from './api.service';
import { PaginatedResponse, DataResponse, FilterParams } from '../models/api-response.model';

export interface Vehiculo {
  id: number;
  placa: string;
  marca: string;
  modelo: string;
  anio: number;
  capacidad_tn: number;
  tipo_vehiculo: string;
  estado: string;
  socio_id: number;
  socio?: { id: number; nombre: string };
}

export interface VehiculoCreate {
  placa: string;
  marca: string;
  modelo: string;
  anio: number;
  capacidad_tn: number;
  tipo_vehiculo: string;
  socio_id: number;
}

export interface VehiculoFilter extends FilterParams {
  socio_id?: number;
}

@Injectable({ providedIn: 'root' })
export class VehiculoService {
  private endpoint = '/vehiculos';

  constructor(private api: ApiService) {}

  listar(params?: VehiculoFilter): Observable<PaginatedResponse<Vehiculo>> {
    return this.api.get<PaginatedResponse<Vehiculo>>(this.endpoint, params);
  }

  obtener(id: number): Observable<Vehiculo> {
    return this.api.get<DataResponse<Vehiculo>>(`${this.endpoint}/${id}`)
      .pipe(map(res => res.data));
  }

  crear(data: VehiculoCreate): Observable<Vehiculo> {
    return this.api.post<DataResponse<Vehiculo>>(this.endpoint, data)
      .pipe(map(res => res.data));
  }

  actualizar(id: number, data: Partial<VehiculoCreate>): Observable<Vehiculo> {
    return this.api.put<DataResponse<Vehiculo>>(`${this.endpoint}/${id}`, data)
      .pipe(map(res => res.data));
  }

  eliminar(id: number): Observable<void> {
    return this.api.delete<void>(`${this.endpoint}/${id}`);
  }

  obtenerDisponibles(): Observable<Vehiculo[]> {
    return this.api.get<DataResponse<Vehiculo[]>>(`${this.endpoint}/disponibles`)
      .pipe(map(res => res.data));
  }
}

File: backend/app/application/services/vehiculos/vehiculo_service.py
"""
Services - Vehiculos
"""
from typing import Optional, List, Dict, Any
from datetime import date

from app.core.unit_of_work import UnitOfWork
from app.domain.exceptions import (
    EntityNotFoundException,
    BusinessRuleException,
    DuplicateEntityException
)


class VehiculoService:
    """Servicio de aplicación para gestión de vehículos"""
    
    def __init__(self, uow: UnitOfWork):
        self.uow = uow
    
    def listar(
        self,
        page: int = 1,
        limit: int = 20,
        estado: Optional[str] = None,
        socio_id: Optional[int] = None
    ) -> Dict[str, Any]:
        """Listar vehículos con filtros"""
        with self.uow:
            filtros = {}
            if estado:
                filtros["estado"] = estado
            if socio_id:
                filtros["socio_id"] = socio_id
            
            vehiculos = self.uow.vehiculos.get_all(
                skip=(page - 1) * limit,
                limit=limit,
                filters=filtros
            )
            total = self.uow.vehiculos.count(filtros)
            
            return {
                "data": [self._to_dict(v) for v in vehiculos],
                "total": total,
                "page": page,
                "limit": limit
            }
    
    def obtener(self, vehiculo_id: int) -> Dict[str, Any]:
        """Obtener detalle de vehículo"""
        with self.uow:
            vehiculo = self.uow.vehiculos.get_with_socio(vehiculo_id)
            if not vehiculo:
                raise EntityNotFoundException("Vehículo", vehiculo_id)
            return self._to_dict(vehiculo, incluir_socio=True)
    
    def crear(self, datos: Dict[str, Any], usuario_id: int) -> Dict[str, Any]:
        """Crear nuevo vehículo"""
        with self.uow:
            # Validar placa única
            if self.uow.vehiculos.get_by_placa(datos["placa"]):
                raise DuplicateEntityException("Vehículo", "placa", datos["placa"])
            
            # Validar socio existe
            socio = self.uow.socios.get_by_id(datos["socio_id"])
            if not socio:
                raise EntityNotFoundException("Socio", datos["socio_id"])
            
            vehiculo = self.uow.vehiculos.create({
                **datos,
                "placa": datos["placa"].upper(),
                "estado": "DISPONIBLE"
            })
            self.uow.commit()
            
            return self._to_dict(vehiculo)
    
    def actualizar(self, vehiculo_id: int, datos: Dict[str, Any]) -> Dict[str, Any]:
        """Actualizar vehículo"""
        with self.uow:
            vehiculo = self.uow.vehiculos.get_by_id(vehiculo_id)
            if not vehiculo:
                raise EntityNotFoundException("Vehículo", vehiculo_id)
            
            # Validar placa única si cambió
            if "placa" in datos and datos["placa"] != vehiculo.placa:
                existente = self.uow.vehiculos.get_by_placa(datos["placa"])
                if existente:
                    raise DuplicateEntityException("Vehículo", "placa", datos["placa"])
            
            self.uow.vehiculos.update(vehiculo_id, datos)
            self.uow.commit()
            
            return self.obtener(vehiculo_id)
    
    def eliminar(self, vehiculo_id: int) -> bool:
        """Eliminar vehículo (soft delete)"""
        with self.uow:
            vehiculo = self.uow.vehiculos.get_by_id(vehiculo_id)
            if not vehiculo:
                raise EntityNotFoundException("Vehículo", vehiculo_id)
            
            if vehiculo.estado == "EN_VIAJE":
                raise BusinessRuleException("No se puede eliminar un vehículo en viaje")
            
            self.uow.vehiculos.delete(vehiculo_id)
            self.uow.commit()
            return True
    
    def obtener_disponibles(self) -> List[Dict[str, Any]]:
        """Obtener vehículos disponibles para asignar"""
        with self.uow:
            vehiculos = self.uow.vehiculos.get_disponibles()
            return [self._to_dict(v) for v in vehiculos]
    
    def _to_dict(self, v, incluir_socio: bool = False) -> Dict[str, Any]:
        """Convertir entidad a diccionario"""
        data = {
            "id": v.id,
            "placa": v.placa,
            "marca": v.marca,
            "modelo": v.modelo,
            "año": v.año,
            "tipo": v.tipo,
            "capacidad_ton": float(v.capacidad_ton) if v.capacidad_ton else None,
            "estado": v.estado
        }
        if incluir_socio and v.socio:
            data["socio"] = {
                "id": v.socio.id,
                "nombre": f"{v.socio.nombre} {v.socio.apellido}"
            }
        return data
